File: app/src/main/java/com/nutomic/syncthingandroid/service/ReceiverManager.java
package dev.benedek.syncthingandroid.service;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.util.Log;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class ReceiverManager {

    private static final String TAG = "ReceiverManager";

    private static List<BroadcastReceiver> mReceivers = new ArrayList<BroadcastReceiver>();

    public static synchronized void registerReceiver(Context context, BroadcastReceiver receiver, IntentFilter intentFilter) {
        mReceivers.add(receiver);
        context.registerReceiver(receiver, intentFilter);
        Log.v(TAG, "Registered receiver: " + receiver + " with filter: " + intentFilter);
    }

    public static synchronized boolean isReceiverRegistered(BroadcastReceiver receiver) {
        return mReceivers.contains(receiver);
    }

    public static synchronized void unregisterAllReceivers(Context context) {
        if (context == null) {
            Log.e(TAG, "unregisterReceiver: context is null");
            return;
        }
        Iterator<BroadcastReceiver> iter = mReceivers.iterator();
        while (iter.hasNext()) {
            BroadcastReceiver receiver = iter.next();
            if (isReceiverRegistered(receiver)) {
                try {
                    context.unregisterReceiver(receiver);
                    Log.v(TAG, "Unregistered receiver: " + receiver);
                } catch(IllegalArgumentException e) {
                    // We have to catch the race condition a registration is still pending in android
                    // according to https://stackoverflow.com/a/3568906
                    Log.w(TAG, "unregisterReceiver(" + receiver + ") threw IllegalArgumentException");
                }
                iter.remove();
            }
        }
    }
}

File: app/src/main/kotlin/dev/benedek/syncthingandroid/service/ReceiverManager.kt
package dev.benedek.syncthingandroid.service

import android.content.BroadcastReceiver
import android.content.Context
import android.content.IntentFilter
import android.util.Log

object ReceiverManager {
    private const val TAG = "ReceiverManager"

    private val mReceivers: MutableList<BroadcastReceiver?> = ArrayList()

    @JvmStatic
    @Synchronized
    fun registerReceiver(
        context: Context,
        receiver: BroadcastReceiver?,
        intentFilter: IntentFilter?
    ) {
        mReceivers.add(receiver)
        // FIXME
        context.registerReceiver(receiver, intentFilter)
        Log.v(TAG, "Registered receiver: $receiver with filter: $intentFilter")
    }

    @Synchronized
    fun isReceiverRegistered(receiver: BroadcastReceiver?): Boolean {
        return mReceivers.contains(receiver)
    }

    @JvmStatic
    @Synchronized
    fun unregisterAllReceivers(context: Context?) {
        if (context == null) {
            Log.e(TAG, "unregisterReceiver: context is null")
            return
        }
        val iter = mReceivers.iterator()
        while (iter.hasNext()) {
            val receiver = iter.next()
            if (isReceiverRegistered(receiver)) {
                try {
                    context.unregisterReceiver(receiver)
                    Log.v(TAG, "Unregistered receiver: $receiver")
                } catch (_: IllegalArgumentException) {
                    // We have to catch the race condition a registration is still pending in android
                    // according to https://stackoverflow.com/a/3568906
                    Log.w(
                        TAG,
                        "unregisterReceiver($receiver) threw IllegalArgumentException"
                    )
                }
                iter.remove()
            }
        }
    }
}
